data-pipeline/src/pipeline/main.py

```diff
@@ -441,20 +463,121 @@ def run_compute_rag(set_type, year):
     return time_taken
 
 
-def handle_msg(worker_client, msg, worker_queue, complete_queue):
-    msg_payload = json.loads(msg.content)
-    try:
-        msg_payload["pre_process_duration"] = pre_process_data(
-            worker_client, msg_payload["type"], msg_payload["year"]
+def run_user_defined_rag(
+    set_type: str,
+    year: int,
+    run_id: str,
+    comparator_set: pd.DataFrame,
+):
+    """
+    Perform user-defined RAG calculations.
+
+    Use the pre-processed "all-schools" data to guarantee coverage of
+    the user-defined comparator-set.
+
+    :param set_type: the triggering message type
+    :param year: financial year in question
+    :param run_id: unique run identifier
+    :param comparator_set: user-defined comparator-set
+    :return: duration of RAG calculations
+    """
+    all_schools = prepare_data(
+        pd.read_parquet(
+            get_blob("pre-processed", f"{set_type}/{year}/all_schools.parquet")
         )
+    )
 
-        msg_payload["comparator_set_duration"] = compute_comparator_sets(
-            msg_payload["type"], msg_payload["year"]
-        )
+    st = time.time()
+    logger.info(f"Computing user-defined RAG ({run_id}).")
+    df = pd.DataFrame(compute_user_defined_rag(all_schools, comparator_set)).set_index(
+        "URN"
+    )
+    logger.info(
+        f"Computing user-defined ({run_id}) RAG. Done in {time.time() - st:.2f} seconds."
+    )
 
-        msg_payload["rag_duration"] = run_compute_rag(
-            msg_payload["type"], msg_payload["year"]
-        )
+    write_blob(
+        "metric-rag",
+        f"{set_type}/{run_id}/user-defined.parquet",
+        df.to_parquet(),
+    )
+
+    insert_metric_rag(
+        run_type=set_type,
+        set_type="user-defined",
+        year=run_id,
+        df=df,
+    )
+
+
+def handle_msg(
+    worker_client: Client,
+    msg: QueueMessage,
+    worker_queue: QueueClient,
+    complete_queue: QueueClient,
+):
+    """
+    Process an incoming message.
+
+    TODO: standard message format?
+
+    User-defined comparator set message _content_ will be of the form:
+
+    ```json
+    {
+        "jobId": "24463424-9642-4314-bb55-45424af6e812",
+        "type": "comparator-set",
+        "runId": "c321ef6a-3b1c-4ce2-8e32-0d0167bf2fa7",
+        "year": 2022,
+        "urn": "106057",
+        "payload": {
+            "kind": "ComparatorSetPayload",
+            "set": [
+                "145799",
+                "142875"
+            ]
+        }
+    }
+    ```
+
+    Note: user-defined comparator sets will assume pre-processing has
```

Comment:
Is this safe to assume? i.e. that the main processing will have already occurred and that a given year's pre-processed data will already exist, ready for user-defined comparator-sets?

Reply:
I think this is a safe assumption to make. Realistically I don't see why we would show a particular "years" data if we did not have that data loaded.

Reply:
Yep.. for the minute this as safe assumption.